### PR TITLE
Add let's encrypt support

### LINKED ISSUE #89: ACME (letsencrypt) interface or plugin
Let's encrypt had proposed an ietf draft to automate ssl certificats creation and renewal. It would be awesome in the futur to have auto-https frontend.

https://github.com/ietf-wg-acme/acme/blob/master/draft-ietf-acme-acme.md

### PR DESCRIPTION
**Pull request by emilevauge**
This PR adds let's encrypt support.

- [x] basic storage file
- [x] on demand certificate
- [x] renew
- [x] proxy challenge
- [x] doc

fixes #89

### CODE REVIEW

**Comment by emilevauge** [comment]
@rebaseTraefik

**Comment by emilevauge** [comment]
@rebaseTraefik

**Comment by vdemeester** [comment]
@emilevauge did a quick code reading and I definitely would prefer for `acme.go` to be isolated in a package if possible (and defining a clean API to use it inside traefik).

File: acme.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
+/*
+Copyright
+*/
+package main
+
+import (
```
Comment (vdemeester) [inline acme.go]: I would :heart: this to be organized :angel:

**Comment by emilevauge** [comment]
Now using `tls-sni-01` challenge instead of `http-01` which only require port `443` ;)
@vdemeester, everythong is in an `acme` package, and you can now use it as a library:

```go
func main(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/hello", helloWorld)
	tlsConfig := tls.Config{}
	server := &http.Server{
		Addr:      ":443",
		Handler:   mux,
		TLSConfig: &tlsConfig,
	}

	acme := ACME{
		Email: "test@traefik.io",
		Domains: []Domain{
			Domain{
				Main: "mydomain.com",
				SANs: []string{"test1.mydomain.commay", "test2.mydomain.commay"},
			},
			Domain{
				Main: "myotherdomain.com",
				SANs: []string{"test1.myotherdomain.commay", "test2.myotherdomain.commay"},
			},
		},
		StorageFile: "acme.json",
		OnDemand:    true,
	}

	err := acme.CreateACMEConfig(&tlsConfig, nil)
	if err != nil {
		log.Fatal("ACME: ", err)
	}
	listener, err := tls.Listen("tcp", ":443", &tlsConfig)
	if err != nil {
		log.Fatal("Listener: ", err)
	}
	server.Serve(listener)
}

func helloWorld(w http.ResponseWriter, req *http.Request) {
	io.WriteString(w, "hello, world!\n")
}
```

File: acme/acme.go [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+			domainsCertificate.Certificate = acmeCert
+			tlsCert, err := tls.X509KeyPair(acmeCert.Certificate, acmeCert.PrivateKey)
+			if err != nil {
+				return err
+			}
+			domainsCertificate.tlsCert = &tlsCert
+			return nil
+		}
+	}
+	return errors.New("Certificate to renew to found from domain " + domain.Main)
```
Comment (vdemeester) [inline acme/acme.go]: I'm not sure I understand this sentence :angel:

File: acme/acme.go [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+	cert := DomainsCertificate{Domains: domain, Certificate: acmeCert, tlsCert: &tlsCert}
+	dc.Certs = append(dc.Certs, &cert)
+	return &cert, nil
+}
+
+func (dc *DomainsCertificates) getCertificateForDomain(domainToFind string) (*DomainsCertificate, bool) {
+	dc.lock.RLock()
+	defer dc.lock.RUnlock()
+	for _, domainsCertificate := range dc.Certs {
+		domains := append([]string{domainsCertificate.Domains.Main}, domainsCertificate.Domains.SANs...)
```
Comment (vdemeester) [inline acme/acme.go]: This is could end up modifying `domainsCertificate.Domains.Main`, is this intended ? or is it a side effect, and the only thing required is to loop over the union of `domainsCertificate.Domains.Main` and `domainsCertificate.Domains.SANs` ?

File: acme/acme.go [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+
+// DomainsCertificate contains a certificate for multiple domains
+type DomainsCertificate struct {
+	Domains     Domain
+	Certificate *Certificate
+	tlsCert     *tls.Certificate
+}
+
+// GetEmail returns email
+func (a Account) GetEmail() string {
```
Comment (vdemeester) [inline acme/acme.go]: These methods would be better put close to the structure definitions (just a nit, but it makes it easier to read / follow)

File: acme/acme.go [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+}
+
+// Domain holds a domain name with SANs
+type Domain struct {
+	Main string
+	SANs []string
+}
+
+// CreateACMEConfig creates a tls.config from using ACME configuration
+func (a *ACME) CreateACMEConfig(tlsConfig *tls.Config, CheckOnDemandDomain func(domain string) bool) error {
```
Comment (vdemeester) [inline acme/acme.go]: s/CreateACMEConfig/CreateConfig/ (ACME is already in the struct name :angel: )

File: acme/acme.go [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+	acme.Logger = fmtlog.New(ioutil.Discard, "", 0)
+	// TODO: generate default cert if empty
```
Comment (vdemeester) [inline acme/acme.go]: Is this still valid ? ('cause I think it's already done : https://github.com/containous/traefik/pull/224/files#diff-faa8f9dd50dafacfcb04a8203ce3deebR172)

File: acme/acme.go (hunk 5 above)
Comment (vdemeester) [inline acme/acme.go]: It's a pretty big function :scream_cat: ; but meh.. we can refactor it afterwards probably

File: acme/acme.go [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+		if err != nil {
+			return err
+		}
+		tlsConfig.Certificates = append(tlsConfig.Certificates, *cert)
+	}
+	var account *Account
+	var needRegister bool
+
+	// if certificates in storage, load them
+	if fileInfo, err := os.Stat(a.StorageFile); err == nil && fileInfo.Size() != 0 {
```
Comment (vdemeester) [inline acme/acme.go]: This part could be split into several methods : `getCertificatesFromStorage` and inside it, if no present, call a `createCertificates` that would do the "Generating ACME account"

File: acme/acme.go [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+	}
+
+	// The client has a URL to the current Let's Encrypt Subscriber
+	// Agreement. The user will need to agree to it.
+	err = client.AgreeToTOS()
+	if err != nil {
+		return err
+	}
+
+	go func() {
```
Comment (vdemeester) [inline acme/acme.go]: I would really prefer to have a `go a.retrieveCertificates(…)` to make the code clearer.

File: acme/acme.go [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+				return nil, nil
+			}
+			return a.loadCertificateOnDemand(client, account, clientHello)
+		}
+		return nil, nil
+	}
+
+	ticker := time.NewTicker(24 * time.Hour)
+	go func() {
+		time.Sleep(24 * time.Hour)
```
Comment (vdemeester) [inline acme/acme.go]: Why is this `Sleep` needed ?

File: acme/acme.go [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+	return nil
+}
+
+func (a *ACME) renewCertificates(client *acme.Client, Account *Account) error {
+	for _, certificateResource := range Account.DomainsCertificate.Certs {
+		// <= 7 days left, renew certificate
+		if certificateResource.tlsCert.Leaf.NotAfter.Before(time.Now().Add(time.Duration(24 * 7 * time.Hour))) {
+			log.Debugf("Renewing certificate %+v", certificateResource.Domains)
+			renewedCert, err := client.RenewCertificate(acme.CertificateResource{
+				Domain:        certificateResource.Certificate.Domain,
```
Comment (vdemeester) [inline acme/acme.go]: Another "nit" : a little `var` to make it simpler ?

```go
certificate := certificateResource.Certificate
// […]
domain: certificate.Domain
```

File: acme/acme.go [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+		}
+	}
+	return nil
+}
+
+func (a *ACME) buildACMEClient(Account *Account) (*acme.Client, error) {
+
+	// A client facilitates communication with the CA server. This CA URL is
+	// configured for a local dev instance of Boulder running in Docker in a VM.
+	caServer := "https://acme-v01.api.letsencrypt.org/directory"
```
Comment (vdemeester) [inline acme/acme.go]: Not sure I'm understanding the comment. And shouldn't be `caServer` configurable ?

File: cmd.go [hunk 12]
```diff
@@ -166,15 +166,15 @@ func init() {
 	traefikCmd.PersistentFlags().StringVar(&arguments.Boltdb.Endpoint, "boltdb.endpoint", "127.0.0.1:4001", "Boltdb server endpoint")
 	traefikCmd.PersistentFlags().StringVar(&arguments.Boltdb.Prefix, "boltdb.prefix", "/traefik", "Prefix used for KV store")
 
-	viper.BindPFlag("configFile", traefikCmd.PersistentFlags().Lookup("configFile"))
-	viper.BindPFlag("graceTimeOut", traefikCmd.PersistentFlags().Lookup("graceTimeOut"))
-	//viper.BindPFlag("defaultEntryPoints", traefikCmd.PersistentFlags().Lookup("defaultEntryPoints"))
-	viper.BindPFlag("logLevel", traefikCmd.PersistentFlags().Lookup("logLevel"))
+	_ = viper.BindPFlag("configFile", traefikCmd.PersistentFlags().Lookup("configFile"))
```
Comment (vdemeester) [inline cmd.go]: Why this changes (and below) ?

File: provider/docker.go [hunk 13]
```diff
@@ -34,35 +34,39 @@ type DockerTLS struct {
...
-		return err
-	}
-	log.Debug("Docker connection established")
-	if provider.Watch {
-		dockerEvents := make(chan *docker.APIEvents)
-		dockerClient.AddEventListener(dockerEvents)
-		log.Debug("Docker listening")
-		go func() {
-			operation := func() error {
+	go func() {
```
Comment (vdemeester) [inline provider/docker.go]: Same here, why this change ? :angel:

File: acme/acme.go (hunk 3 above)
Comment (emilevauge) [inline acme/acme.go]: I just want to loop on the union. How could this modify `domainsCertificate.Domains.Main`?

**Comment by vdemeester** [comment]
Works well ! Looks good, although I have a few remarks, questions and nit !!
But I can't wait to get this merged though ! :wink:

File: acme/acme.go (hunk 7 above)
Comment (emilevauge) [inline acme/acme.go]: Hum, I prefer that `getCertificatesFromStorage` stays away from generating ACME account ;)

File: acme/acme.go (hunk 8 above)
Comment (emilevauge) [inline acme/acme.go]: i agree

File: acme/acme.go (hunk 9 above)
Comment (emilevauge) [inline acme/acme.go]: Only to postpone the renewing check later ;)

File: acme/acme.go (hunk 10 above)
Comment (emilevauge) [inline acme/acme.go]: Hum, sorry, could you elaborate? I don't get it ^^

File: cmd.go (hunk 12 above)
Comment (emilevauge) [inline cmd.go]: Just some cleaning, and changed `MaxIdleConnsPerHost` default value.

File: provider/docker.go (hunk 13 above)
Comment (emilevauge) [inline provider/docker.go]: In case of docker connection failing, now there is a retry exponential backoff while reconnecting ;)

File: acme/acme.go (hunk 3 above)
Comment (vdemeester) [inline acme/acme.go]: `domainsCertificate.Domains.Main` is a slice right ? Then, `append` checks if the slice needs to grow or not. If it does need to grow, it will allocate a new array and `domainsCertificate.Domains.Main` is not modified (everything is copied/moved) to the new array. But if it does not need to grow, it will just add the element to the current array, and thus, `domainsCertificate.Domains.Main` is modified.

Something like that :

```go
a := []string{ "Janvier", "Fevrier", "Mars" }
b := append(months, "Avril")
c := append(months, "Mai")

fmt.Println(len(a), cap(a)) // "3, 3"
fmt.Printf("%q\n", a)       // ["Janvier" "Février" "Mars"]
fmt.Println(len(b), cap(b)) // "4 6"
fmt.Printf("%q\n", b)       // ["Janvier" "Février" "Mars" "Avril"]
fmt.Println(len(c), cap(c)) // "5 6"
fmt.Printf("%q\n", c)       // ["Janvier" "Février" "Mars" "Avril" "Mai"]

fmt.Printf("%q\n", b[:5])   // ["Janvier" "Février" "Mars" "Avril" "Mai"]
// … mais pas a
fmt.Printf("%q\n", a[:5])   // panic : out of range
```

I think it would be better create a new slice and put the elements of both inside it :angel:.

File: acme/acme.go (hunk 10 above)
Comment (vdemeester) [inline acme/acme.go]: It's just long to write and read `certificateResource.Certificate.Domain`, but meh :smile:

File: cmd.go (hunk 12 above)
Comment (vdemeester) [inline cmd.go]: yeah, but why the `_ =` ?

File: provider/docker.go (hunk 13 above)
Comment (vdemeester) [inline provider/docker.go]: hum ok, I'll have to check that but as long as it works, I'm fine with it :sweat_smile:

File: acme/acme.go [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+				return nil, nil
+			}
+			return a.loadCertificateOnDemand(client, account, clientHello)
+		}
+		return nil, nil
+	}
+
+	ticker := time.NewTicker(24 * time.Hour)
+	go func() {
+		time.Sleep(1 * time.Hour)
```
Comment (vdemeester) [inline acme/acme.go]: Hum, this still feels weird to me. You sleep one hour before the loop, but in the loop, you wait for the ticker for 24 hours (that will have been already started), so in the end you waited for 24 hours, with or without this Sleep. Am I wrong on that ? (maybe my assumption on how Ticker works is wrong though :smile:)

File: acme/acme.go (hunk 3 above)
Comment (emilevauge) [inline acme/acme.go]: Ops :angel:

File: cmd.go (hunk 12 above)
Comment (emilevauge) [inline cmd.go]: Hum, because of errcheck :angel: 
I just don't want to check error in that case...

Comment (vdemeester) [inline cmd.go]: aahhh I see :+1:

File: acme/acme.go (hunk 14 above)
Comment (emilevauge) [inline acme/acme.go]: Hum, I thought ticker would begin ASAP, but it's not ;)

**Comment by emilevauge** [comment]
@vdemeester thanks for your great review!
Fixed!

**Comment by vdemeester** [comment]
And it's :green_heart: !
LGTM 🐯
